java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java

```diff
@@ -0,0 +1,286 @@
+      }
+      return substitutedType;
+    }
+    return arrayType;
+  }
+
+  TypeSubstitution getSubstitutionFromTypeParams(List<JavaType.TypeVariableJavaType> typeVariableTypes, List<JavaType> typeParams) {
+    TypeSubstitution substitution = new TypeSubstitution();
+    if (typeVariableTypes.size() == typeParams.size()) {
+      // create naive substitution
+      for (int i = 0; i < typeVariableTypes.size(); i++) {
+        JavaType.TypeVariableJavaType typeVariableType = typeVariableTypes.get(i);
+        JavaType typeParam = typeParams.get(i);
+        substitution.add(typeVariableType, typeParam);
+      }
+    }
+    return substitution;
+  }
+
+  private TypeSubstitution getSubstitutionFromArguments(JavaSymbol.MethodJavaSymbol method, List<JavaType> formals, List<JavaType> argTypes) {
+    TypeSubstitution substitution = new TypeSubstitution();
+
+    if (formals.size() > argTypes.size() || formals.isEmpty() || argTypes.isEmpty()) {
+      // no need to try to infer types
+      return substitution;
+    }
+
+    List<JavaType.TypeVariableJavaType> typeVariablesToInfer = new ArrayList<>(method.typeVariableTypes);
+    // compete with type variable from the scope
+    for (JavaType argType : argTypes) {
+      JavaSymbol owner = argType.symbol.owner;
+      if (owner.isMethodSymbol() && ((JavaSymbol.MethodJavaSymbol) owner).typeVariableTypes.contains(argType)) {
+        typeVariablesToInfer.add((JavaType.TypeVariableJavaType) argType);
+      }
+    }
+
+    if (!typeVariablesToInfer.isEmpty()) {
+      for (int i = 0; i < formals.size(); i++) {
+        JavaType formalType = formals.get(i);
+        JavaType argType = argTypes.get(i);
+
+        if (formalType.isTagged(JavaType.TYPEVAR)) {
+          completeSubstitution(substitution, formalType, argType);
+        } else if (formalType.isTagged(JavaType.ARRAY) && argType.isTagged(JavaType.ARRAY)) {
+          completeSubstitution(substitution, ((JavaType.ArrayJavaType) formalType).elementType, ((JavaType.ArrayJavaType) argType).elementType);
+        } else if (formalType.isTagged(JavaType.ARRAY) && method.isVarArgs() && i == formals.size() - 1) {
+          completeSubstitution(substitution, ((JavaType.ArrayJavaType) formalType).elementType, argType);
+        } else if (isParametrizedType(formalType) && isParametrizedType(argType)) {
+          List<JavaType> formalTypeSubstitutedTypes = ((JavaType.ParametrizedTypeJavaType) formalType).typeSubstitution.substitutedTypes();
+          List<JavaType> argTypeSubstitutedTypes = ((JavaType.ParametrizedTypeJavaType) argType).typeSubstitution.substitutedTypes();
+          TypeSubstitution newSubstitution = getSubstitutionFromArguments(method, formalTypeSubstitutedTypes, argTypeSubstitutedTypes);
+          substitution = mergeTypeSubstitutions(substitution, newSubstitution);
+        } else if (isParametrizedType(formalType) && !isParametrizedType(argType) && formalType.erasure() == argType) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)